src/main/java/org/prebid/server/bidder/ttx/TtxBidder.java

```diff
@@ -1,64 +1,199 @@
+        }
+
+        return HttpRequest.<BidRequest>builder()
+                .method(HttpMethod.POST)
+                .uri(endpointUrl)
+                .headers(HttpUtil.headers())
+                .payload(modifiedRequest)
+                .body(mapper.encode(modifiedRequest))
                 .build();
     }
 
+    private void modifyRequest(BidRequest request,
+                               BidRequest.BidRequestBuilder modifiedRequest,
+                               List<BidderError> errors) {
+        final Imp firstImp = request.getImp().get(0);
+        final Imp.ImpBuilder modifiedFirstImp = firstImp.toBuilder();
+        try {
+            processModify(request, modifiedRequest, firstImp, modifiedFirstImp);
+        } catch (PreBidException e) {
+            errors.add(BidderError.badInput(e.getMessage()));
+        }
+
+        final List<Imp> imps = new ArrayList<>(request.getImp());
+        imps.set(0, modifiedFirstImp.build());
+        modifiedRequest.imp(imps);
+    }
+
+    private void processModify(BidRequest request,
+                               BidRequest.BidRequestBuilder modifiedRequest,
+                               Imp firstImp,
+                               Imp.ImpBuilder modifiedFirstImp) {
+        final ExtImpTtx extImpTtx;
+        extImpTtx = parseImpExt(firstImp);
+        final String zoneId = extImpTtx.getZoneId();
+        final TtxImpExt ttxImpExt = TtxImpExt.of(
+                TtxImpExtTtx.of(extImpTtx.getProductId(), StringUtils.isNotBlank(zoneId) ? zoneId : null));
+
+        modifiedFirstImp.ext(mapper.mapper().valueToTree(ttxImpExt)).build();
+        final Site site = request.getSite();
+        final Site.SiteBuilder siteBuilder = site == null ? Site.builder() : site.toBuilder();
+        modifiedRequest.site(siteBuilder.id(extImpTtx.getSiteId()).build());
+        final Video video = firstImp.getVideo();
+        if (firstImp.getVideo() != null) {
+            modifiedFirstImp.video(modifyVideo(video, extImpTtx.getProductId()));
+        }
+    }
+
+    private ExtImpTtx parseImpExt(Imp imp) {
+        try {
+            return mapper.mapper().convertValue(imp.getExt(), TTX_EXT_TYPE_REFERENCE).getBidder();
+        } catch (IllegalArgumentException e) {
+            throw new PreBidException(e.getMessage());
+        }
+    }
+
+    private Video modifyVideo(Video video, String prod) {
+        if (video.getW() == null
+                || video.getW() == 0
+                || video.getH() == null
+                || video.getH() == 0
+                || CollectionUtils.isEmpty(video.getProtocols())
+                || CollectionUtils.isEmpty(video.getMimes())
+                || CollectionUtils.isEmpty(video.getPlaybackmethod())
+        ) {
+            throw new PreBidException("One or more invalid or missing video field(s) "
+                    + "w, h, protocols, mimes, playbackmethod");
+        }
+        final Video.VideoBuilder modifiedVideo = video.toBuilder();
```

Comment:
Is there the case when we don't need to update the video, so creation of builder is redundant.
It is not so critical for bidders, but in core we use approach as in `org.prebid.server.auction.AuctionRequestFactory#fillImplicitParameters` method, like extracting all needed fields and then if at least one them is desired to be updated - create the builder and set fields.

Again, not critical, but just to think.